User rating for video items

razzeee opened the pull request:
So this is not about an overall rating like we already have, but the rating the user gave this item. So this enables the user to save an integer on the musicvideo, movie, episode and tvshow items. 

First thing to discuss might be if it should be an integer and if we should limit it. I for myself am liking integers most preferably 1-10. Because that's like the normal rating field, which also goes from 1-10.

Then I have some TODOs still in the commit where I'm not so sure how to handle this perfectly. Please have a look.
I'm also not sure about https://github.com/xbmc/xbmc/pull/7561/files#diff-dbf978b6cac348a2af93d3beec420e43R4571 as I would have thought that I'm not the first dev that needs to add something to the movies table and this code should probably already exist somewhere?

I've basically did this:
- Added user rating to musicvideo, movie, episode and tvshow databases
- Made user rating it scrapable
- Added user rating to json rpc (get/set)
- Added user rating to videoplayer and listitem for skinners
- Prepared a user rating sort option

I've tested setting and getting via json rpc, displaying it via listitem and scraping a movie on windows. That's working for me, but this will still need some heavy reviewing.
Hope I'm not the only one who sees value in having a rating in the video library.

razzeee committed:
Added user rating to musicvideo, movie, episode and tvshow databases
Made user rating it scrapable

phil65 commented:
Nice. I see value in having that, too.
@Montellese probably for JSON part and @mkortstiege for db stuff.

MartijnKaijser commented:
As a suggestion would it make sense to do the same as the art table where we could freely insert additional "ratings" with some initial  fixed names like for userrating?
Then we could add additional ratings per provider which could be displayed by the skin or even averaged. 

If that's out of capabilities very nice addition :) 

I'd  say int 1-10 or 0-10

Tolriq commented:
1-10 for value, 0 for no value, else it's impossible for skin / remote to know if it's a 0 rating or a missing value and decide to switch to scraped rating. (Or add a -1 but that would not be consistent with the rest of the apis).

What about the audio rating that is AFAIK a 0-5 value ? Should not the user defined rating a common range for all medias ?

But nice addition and often asked :)

Razzeee commented:
@phil65 
I've seen that part and was puzzled what it does, as it's not giving access to all fields. But I think I can add that part very easily.

@MartijnKaijser 
Well I will have to look at multiple values in one field for one of the things I would also like to add someday. But I think it has no value when coming to user ratings, why should I think a movie is a 7 on tmdb and think its a 4 on trakt?

Razzeee commented:
Also to everybody reading this, what's up with the 'idSet' field? It's a really tripping stone when adding a fileld or it the mapper (cXX) thing depricated and shouldn't be used? In my opinion it should be moved to the start of the table.

Razzeee commented:
@phil65 
can you name a use case for setting this via skin? I'm not sure where a skin would get rating data from?
Come to think of it, I think there also is no valid usecase for this beeing scrapable, as I'm not aware of scrapers with credentials/auth so it can get the users rating. Please correct me.

MartijnKaijser commented:
@Razzeee not sure what you mean with the comment towards phill65.
For music we can set the rating ourselves. TMDB/Trakt allows you to have your own rating stored which in turn you could sync between local install/provider. True conventional scrapers don't support this however services/python scrapers could take this over in the end. (hope this is what you asked to clarify?)

Razzeee commented:
@MartijnKaijser 
Now I got the Skin Part, thanks. 
Yes, services/python scrapers could do this, but they will use json-rpc not the scraping xml, correct? I'm just thinking the scraping xml part shouldn't be in there.

I will tackle music next then, might be more complicated, as it hands out images as far as I have seen?

@Tolriq  
What do you think, can you handle 10 stars in the yatze frontend? Might get pretty crowded?

MartijnKaijser commented:
@Razzeee yes it should be JSON-RPC for syncing. So user set his rating and it could get synced with upnp/json-rpc to other clients/services.

You could also show 5 stars with half a star interval

Tolriq commented:
Well anything can be handled, it's just that it needs special handling depending on media types and may confuse users to sometimes have to set a rating /5 and other times /10. 

At first I had normalized values in Yatse to 10 star but users where lost about that and told me I want 5 star for music so adapted, and handled setting song rating on a 5 star input.
So not sure about changing audio rating to 0-10 as users are used to 0-5.

For video it's a mess as it depends on scraper, I do use french scraper allociné and rating are based on 5 stars. I think it was normalized to a 0-10 value during scraping but not even sure as I switched to imdb ratings.

But having something stable over everything (scraper included) may help for future.

zag2me commented:
Very nice new feature thanks! Its a shame we have a different rating system for music tracks and albums but maybe that could be changed to 10 in the future as well. So +1 to this 1-10 system and 0 for unrated.

Montellese commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -4043,6 +4046,13 @@ std::string CGUIInfoManager::GetVideoLabel(int item)
+    {
+      std::string strUserRating;
+      if (m_currentFile->GetVideoInfoTag()->m_iUserRating > 0)
+        strUserRating = StringUtils::Format("%i", m_currentFile->GetVideoInfoTag()->m_iUserRating);;
```
Duplicate `;` (same below)

Razzeee replied:
The whole file is crawling with them, not sure if it's okay if I fix them inside this PR? Left them out on purpose, but must have copied somewhere wrong.

MartijnKaijser replied:
should clean that up in separate commit. i could do that for you if you want

Razzeee replied:
@MartijnKaijser 
please go ahead. I'll remove just the two I introduced.

Montellese commented:
The reason for the `idSet` column in the movie table is that we wanted to move away from the `cXX` names because they make queries very hard to read and `idSet` was a property only used by movies so it didn't make any sense to add an extra column to all other tables. I suggest to do the same here i.e. add a `userrating` column at the end of all tables and manually handle reading/setting it. That also makes the update logic much easier.

I would also keep the scraper/XML integration because it's not only for scrapers but also for NFOs so users can set their ratings in there.

Last but not least it would be nice to have GUI integration to be able to rate (watched) stuff (either from the library or optionally after having watched something) but that can also be done in a separate PR.

Razzeee commented:
@Montellese 
I'm really glad to see that we're moving away from that generic structure. The only thing that kept me, was that I thought that I was doing something wrong. So maybe we should add an inline comment there to point readers in the direction of the prefered storage approach?

Haven't thought about NFOs but that makes sounds really useful for some users.

Your probably talking about native kodi gui? Skin guis, should be able to use this with the latest commit (needs testing). The native gui stuff is probably to complicated for me, as I'm still learning and I really hate doing gui even in my "primary" programming language.

Razzeee commented:
@all
do we agree that this feature makes sense for musicvideos, movies, tvshows and episodes? Or is there one that should not be in there?

Razzeee commented:
@Montellese 
Doesn't seem like it's so easy to add the reading/writing manually, probably need to write quiet some code for that as it seems this would be the first non foreign key field with a custom name. For e.g. ```CVideoDatabase::GetDetailsForMovie``` seems like it can only handle column indicies to pull out data? Am I missing something?

Montellese commented:
@Razzeee you still define a column index like `VIDEODB_DETAILS_MOVIE_SET_ID` (beware that this index is on the `movie_view` and not the `movie` table) and use that.

Razzeee commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -2097,15 +2097,19 @@ int CVideoDatabase::SetDetailsForMovie(const std::string& strFilenameAndPath, co
     // and insert the new row
-    std::string sql = "update movie set " + GetValueString(details, VIDEODB_ID_MIN, VIDEODB_ID_MAX, DbMovieOffsets);
+    std::string sql = "UPDATE movie SET " + GetValueString(details, VIDEODB_ID_MIN, VIDEODB_ID_MAX, DbMovieOffsets);
     if (idSet > 0)
```
I tried to do these four lines in a seperate method and just reuse it to construct the sql, but couldn't make it work :/

Montellese replied:
I don't follow.

Razzeee replied:
I tried to do something like:

```
std::string CVideoDatabase::AddUpdateParameter(std::string name, int value)
{
  if (value > 0)
        return PrepareSQL(", %s = %i", name, value);
      else
        return StringUtils::Format(", %s = NULL", name);
}
```
And then calling doing ```sql += AddUpdateParameter("idSet", idSet);``` or sql += AddUpdateParameter("userrating", details->m_iUserRating);

But for some reason the call gets lost in the PrepareSQL part :(

Montellese replied:
`PrepareSQL`, `StringUtils::Format` and `CLog::Log` can't handle `std::string`s but only `(const) char*`s so you would have to use
```
std::string CVideoDatabase::AddUpdateParameter(std::string name, int value)
{
  if (value > 0)
        return PrepareSQL(", %s = %i", name.c_str(), value);
      else
        return StringUtils::Format(", %s = NULL", name.c_str());
}
```

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -2031,7 +2031,7 @@ int CVideoDatabase::SetDetailsForMovie(const std::string& strFilenameAndPath, co
   {
     BeginTransaction();
-
+    
```
There seem to be quite a few whitespace changes (in the wrong direction).

Razzeee replied:
That wasn't meant to slip in there, probably leftovers from trying to trace the problem I mentioned above

Razzeee commented:
@phil65 
Setting user rating from skin should work as of https://github.com/Razzeee/xbmc/commit/4aa1376119d363e8f7f52d2770a957f4429f0ba0 can you test this?

uNiversaI commented:
@Razzeee would be nice if this supported

IncreaseRating / DecreaseRating action IDS http://kodi.wiki/view/Action_IDs currently used for rating music tracks see https://github.com/xbmc/xbmc/pull/6744 for mappings

tamland commented:
How about using a normalized [0, 1] floating point scale? (at least internally and in APIs) That way it's easier to convert between different scales, and you don't loose information when converting from a higher precision scale than 1-10, e.g 1-100.

Razzeee commented:
@uNiversaI 
looking into it.

@tamland 
as long as internally doesn't mean in the database, then that should be fine

Razzeee commented:
@uNiversaI 
IncreaseRating / DecreaseRating added in the latest commit, will need heavy testing. Btw have never used it what happens if you haven't rated that item before?

uNiversaI commented:
@Razzeee For music items you can just increasing / decrease rating till your hearts content,

Currently confluence rating for Music doesn't have out of 10 star rating images (its out of 10 in text for videos) it only uses full star rating images (where music is concerned) I guess to use the star rating you need to add half a star.

Not sure if thats pertinent and or if you want to add star rating via images as well, it would be easy to do the half stars but someone need to do the skinning part.

Razzeee commented:
@tamland 
I have thought a little and I don't think normalizing it should be inside the scope of this and probably should be done by data providers. Otherwise we would also need data on the range.
For e.g.:
If the rating is 78 on a 1-100 scale we need to know that it is a 1-100 scale otherwise we can't normalize it correctly. In theory it could also be a 1-200 scale and you end up with a much different normalized rating.

tamland commented:
@Razzeee I think you misunderstood me. You already do normalize to the 1-10. Just do the same but for 0-1. That means change the db to store float as the other 'rating'. Everyone using the API must convert from/to their scale (just as now). Then for the info label, it can be convert to a presentable format such as 1-10 (or whatever makes sense).

Razzeee commented:
@tamland 
Well I think I left the normalize part out of this as of now.

Montellese commented:
@tamland: why should 0-1 be better than 0-10? We will never show the 0-1 value in the user interface but we will very likely show the 0-10 value so that means that with 0-1 we always have to do the conversion. We can also easily store the 0-10 value as a float if you are afraid of loosing precision when normalizing from a bigger range. IIRC we also store the existing rating for videos as a float in the 0-10 range (which is how it is provided by IMDb and TMDb).

Razzeee commented:
@Montellese 
well on the existing rating, it makes even more sense, as its an average of XXXX user ratings, that gets scraped there. You won't end up with a round value on that.

tamland commented:
@Montellese When float, it isn't. But this one is currently 1-10, i.e. yet another. I think it should at least follow the existing rating field.

Razzeee commented:
I'm trying to get this squashed and rebased today.

Not sure what else should be done, besides more reviewing.
The int vs float thing is probably something the core team has to decide upon. I think it doesn't make a difference and I think ```int``` is easier. Adding it via json isn't a problem as every addon can normalise and round on it's own and the json-rpc is not accepting any floats.
The scraper will just cut off everything after the point, when he gets handed a float. So we end up with 8 when the scraper gets a 8.6
Not ideal, but I also don't know if scrapers are able to round values?

Frontend for confluence should be another PR (not done by me) and I think this could already be used in custom skins. We just need some addons to support this, which I will add to ```script.trakt``` soon.

Razzeee commented:
Squashed and rebased.

da-anda commented:
nice addition. I haven't checked your implementation, but I only wanted to ask if it would make sense to make it multi-user compatible. With that I mean that we support profiles that can share a common library and I think it would be good if the rating could be per profile/user. This doesn't have to be done in this PR, but maybe just be considered in the implementation.

Razzeee commented:
@da-anda 
I thought of that too, but beeing new to the kodi source, I didn't know how you guys handle the profiles right now. Haven't researched that so far, I only know that my previous experience with kodi profiles were quiet bad, so I'm really looking forward to the GSoC 2015 (re-)work.

Razzeee commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -9293,3 +9324,32 @@ bool CVideoDatabase::GetFilter(CDbUrl &videoUrl, Filter &filter, SortDescription
+
+    std::string sql;
+    if (mediaType == MediaTypeMovie)
+      sql = PrepareSQL("UPDATE movie SET userrating=%i WHERE c%02d = %i", rating, VIDEODB_DETAILS_MOVIE_PATH, filePath);
```
This will not work, we need a new solution here

Razzeee commented:
Here is a very basic version of the trakt script that can write movie ratings to kodi when you sync. Be aware that you need this branch build, the setting in trakt activated and there are bugs.
https://github.com/Razzeee/script.trakt/tree/User-Ratings

Montellese commented on xbmc/Application.cpp:
```diff
@@ -2266,6 +2266,39 @@ bool CApplication::OnAction(const CAction &action)
+        CVideoDatabase db;
+        if (db.Open())
+        {
+          db.SetVideoUserRating(m_itemCurrentFile->GetPath(), m_itemCurrentFile->GetVideoInfoTag()->m_iUserRating, m_itemCurrentFile->GetVideoInfoTag()->m_type);
```
Pass in the database id which is available at `CVideoInfoTag::m_iDbId` instead of the path.

mkortstiege replied:
Passing `CVideoInfoTag` would be enough as well as all information are in there.

Razzeee replied:
Yep, will do wasn't sure that it would be unique as I hadn't figured out the m_type part back then.

Razzeee replied:
@mkortstiege 
wouldn't that be quiet an overhead? We just need these three fields afterall

Razzeee commented:
Movies is working nicely so far and I would even say it's done. Will try to test tvshows and episodes in the next days.

Razzeee commented:
Have also tested tvshow and episode user ratings. Working and displaying just fine. I'm done, should be mergable after some more reviews.

mkortstiege commented:
Please don't forget to bump json-rpc as well.

Razzeee commented:
@mkortstiege 
hope it's correct now

MartijnKaijser commented:
@Razzeee you bumped the python version but you need to do JSON and GUI

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4566,6 +4589,13 @@ void CVideoDatabase::UpdateTables(int iVersion)
       }
     }
   }
+  if (iVersion < 94)
```
You forgot to bump the version in `GetSchemaVersion()`.

Razzeee replied:
Does it really need to be one bigger? It's working fine this way. Is there another reason?

Montellese replied:
If you change something in the database code that either requires manual update code or re-creation of the views, triggers, indices, ... you always need to bump the schema version. It already is at `94` so not sure how this is working for you (maybe because you directly updated from `93` or from no previous version at all). Anyone that already has a `MyVideos94.db` won't be getting these manual update instructions and than all queries to these tables will break.

You need to use `95` and also bump the schema version to that.

Razzeee replied:
So basically it's always one bigger then the latest update instructions go. Seems like its mostly important for the alpha users?

Montellese replied:
No that's not 100% correct. If you change a `CREATE VIEW` statement there will be no update instructions but you still need to bump the version so that the views are dropped and re-created during the update procedure.

It's for alpha users but most importantly for developers which run through all kinds of versions of the source code and therefore also the database layout. Furthermore these update procedures are also processed one by one when updating from one stable version to the next up to the version specified in `GetSchemaVersion()`.

Razzeee replied:
Yeah, I meant devs with alpha users :)
Forgot about the indices and statistics for a moment. And I was aware of every iteration of the update logic being important, not sure what I was thinking.
Thank you for explaining!

Razzeee commented:
@MartijnKaijser 
I also bumped the python version correct. There was a comment by tamland I think about that beeing needed? Can't find it right now, probably due to me rewriting the PR history.

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -9293,3 +9324,31 @@ bool CVideoDatabase::GetFilter(CDbUrl &videoUrl, Filter &filter, SortDescription
   return true;
 }
+
+bool CVideoDatabase::SetVideoUserRating(const int& dbId, const int& rating, const MediaType& mediaType)
```
We don't pass primitive data types by reference-to-const.

Razzeee replied:
Please review how it is now.

Montellese replied:
It's ok now, thanks.

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -9293,3 +9324,31 @@ bool CVideoDatabase::GetFilter(CDbUrl &videoUrl, Filter &filter, SortDescription
+  }
+  catch (...)
+  {
+    CLog::Log(LOGERROR, "%s (%i,%i) failed", __FUNCTION__, dbId, rating);
```
Please also log the media type.

Razzeee replied:
Should be fine now

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -9293,3 +9324,31 @@ bool CVideoDatabase::GetFilter(CDbUrl &videoUrl, Filter &filter, SortDescription
+    std::string sql;
+    if (mediaType == MediaTypeMovie)
+      sql = PrepareSQL("UPDATE movie SET userrating=%i WHERE idMovie = %i", rating, dbId);
+    else if (mediaType == MediaTypeEpisode)
```
What about tvshows?

Razzeee replied:
Was under the impression the Increase/Decrease Rating is only there when you watch something. Not sure if that is the case, anyway I added the tvshow code.

Montellese commented on xbmc/video/VideoDatabase.h:
```diff
@@ -300,7 +304,7 @@ const struct SDbTableOffsets DbEpisodeOffsets[] =
   { VIDEODB_TYPE_STRING, my_offsetof(CVideoInfoTag,m_basePath) },
   { VIDEODB_TYPE_INT, my_offsetof(CVideoInfoTag,m_parentPathID) },
-  { VIDEODB_TYPE_STRING, my_offsetof(CVideoInfoTag,m_strUniqueId) }
+  { VIDEODB_TYPE_STRING, my_offsetof(CVideoInfoTag,m_strUniqueId) },
```
Not necessary.

Razzeee replied:
Removed

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -591,6 +598,11 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
   }
+  const TiXmlElement* urElement = movie->FirstChildElement("userrating");
+  if (urElement && (urElement->QueryIntAttribute("max", &max_value) == TIXML_SUCCESS) && max_value >= 1)
+  {
```
No need for the curly braces (yes I know they are also used above but we don't do that anymore).

Razzeee replied:
Removed

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -591,6 +598,11 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
   {
     m_fRating = m_fRating / max_value * 10; // Normalise the Movie Rating to between 1 and 10
   }
+  const TiXmlElement* urElement = movie->FirstChildElement("userrating");
```
You need to reset the value of `max_value` because it might have been changed by the `<rating max="...">` tag parsing.

Razzeee replied:
Not sure about this, it should be set again by my tag parsing. Not seeing the error here right now.

Montellese replied:
Is the `max` attribute mandatory? If so then everything is fine.

Razzeee replied:
No it's not. Because we get the value for userrating initially a few lines before this. But to do the normalise and enter this code block, the max is needed I think? So this will not normalise to ten when there is no max given.
The documentation for ```QueryIntArttribute``` says 
```If the attribute does not exist, then TIXML_NO_ATTRIBUTE is returned.``` this will lead to the if getting ignored

Montellese replied:
Ah you're right. The check against `== TIXML_SUCCESS` will fail and the logic won't be executed. Nevermind then.

Razzeee replied:
We could do it for readability reasons?

Montellese commented:
Commented on a few (minor) issues. Could you please split the commit into separate commits for
* changes to `CVideoDatabase` and `CVideoInfoTag`
* changes to the UI integration
* changes to the python API
* changes to the JSON-RPC API
That makes it easier to debug / bisect specific issues since these changes are pretty much independent (except that all depend on the first change).

Apart from that I guess we need to decide whether we want the user rating to be stored as an integer or a float.

razzeee committed:
Added user rating to videoplayer and listitem for skinners
Added label formatter for user rating
Added IncreaseRating / DecreaseRating for video user ratings
Added sorting option for user rating and show it in tvshows, episode and movie views

razzeee committed:
Added user rating to setInfo on ListItem

razzeee committed:
Added user rating to json rpc for movie, tvshow, episode and musicvideo (get/set)

Razzeee commented:
@Montellese 
Like this?

Montellese commented:
@Razzeee: Thanks, much better.

Razzeee commented:
Needs unittest

uNiversaI commented:
Needs skin implementation or no?

Razzeee commented:
It's working as it is. We may be able to do some nice things in the skin layer but thats not needed or in the scope of this PR. I think.

razzeee committed:
tests: Added unit tests for user ratings (videodb)

Razzeee commented:
Tests added.

phate89 commented:
@Razzeee are you thinking to expand ratings to allow scrapers to scrape more than one rating after this? I was thinking to look into it (but I didn't yet so I have no idea if it's too hard for my kodi knowledge) but if you're already working on it I don't even start.

Razzeee commented:
@phate89 
I will likely work on saving multiple external ids next (currently working on dateAdded for music album and artist, with mixed results)
Go ahead, you should probably have a look on how its done with the thumbs.

Montellese commented:
Uhm is the plan to add multiple user ratings? If so it would be better to not store it directly in the respective SQL table but in a separate table and use a link table. But that makes things quite a bit more complicated.
I just want to avoid adding a column to the `movie` table etc and then having to remove it again in the foreseeable future.

But I thought someone mentioned earlier that multiple user ratings doesn't make any sense?

Razzeee commented:
@Montellese 
I said that, and I think @phate89 is refering to the ratings. Not the user ratings.

Montellese commented:
Ah right, just wanted to be sure.

Montellese commented:
Code looks ok. I guess the only remaining question is whether the user rating should be stored as an integer or a float/double. Personally I think 0-10 in integer is enough for a user rating as nobody gives a rating of 6.72. Those kind of ratings make sense for metadata sites which calculate the average of thousands of user ratings.

Razzeee commented:
@Montellese 
+1
That's exactly my reasoning why I choose int for it.

Tolriq commented:
What about consistency for other consumers of the data ?
Consistency storage does not implies that users can store 6.72 this can be handled at interface level.

Razzeee commented:
@Tolriq 
not sure what you're trying to say

Tolriq commented:
Well those values will be returned to JSON and can be consumed by lot's of things, so having a mix of types (double / int) representing the same kind of values is not very consistent.

(It's more or less the same problem about 10 or 5 max value).

Razzeee commented:
Well to be fair, its not the same type of value. Yeah, they might both have ratings in the name, but they're really really different in useage. And if your refereing to the music rating, I think that one is saved as a char :/

Tolriq commented:
Music is another problem :(

But for video I suppose the most common scenario for users will be display my rating is present or fallback to scraped one.

From my point of view (remote on limited devices) I'd prefer for this scenario to only store 1 rating and do the work during sync for all instead of having 2 columns meaning more space on device + more complex queries = more cpu / battery and so on.

And 1 column also works for the 2 other scenario (display user rating or display scraper rating).

Razzeee commented:
Well, I don't like the idea of combining them, as I as a user would want to see both. Otherwise it could get pretty irritating, when a user isn't sure if this is a scraped rating or one he has given...

MartijnKaijser commented:
agree with @Razzeee . I want MY rating and *others and certainly not combined or as backup.

MartijnKaijser commented on xbmc/SortFileItem.h:
```diff
@@ -39,6 +39,7 @@ typedef enum {
   SORT_METHOD_COUNTRY,
   SORT_METHOD_YEAR,
   SORT_METHOD_VIDEO_RATING,
+  SORT_METHOD_VIDEO_USER_RATING,
```
please also add to
https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/legacy/ModuleXbmcplugin.h#L193

Razzeee replied:
Added, please check if everything is cool.

MartijnKaijser replied:
please add it as same order as you did else where (and it needs to be squashed in the other commit)

razzeee committed:
Added SORT_METHOD_VIDEO_USER_RATING to legacy code

hudokkow commented:
jenkins build this please

MartijnKaijser commented:
needs a rebase

Razzeee commented:
Rebased

Razzeee commented:
Will this be merged in this merge window?

MartijnKaijser commented:
hopefully yes :)
needs a rebase and like to get final ok from @Montellese / @mkortstiege

Razzeee commented:
Rebased

Montellese commented:
Ok from my side.

mkortstiege commented:
jenkins build and merge

mkortstiege commented:
jenkins build and merge (again please).

Razzeee commented:
This might also need an api change label, if I understood correctly

MilhouseVH commented:
This seems to have broken a number of nodes in Confluence, does Confluence need an update, or is this behaviour unexpected?

I can reproduce the issues discussed below using a plain Ubuntu 15.04 build. Reverting this PR resolves all issues.

With this PR, the "Recently added movies" widget is now displayed thus:
![screenshot](http://i.imgur.com/9P677w3.png)

Trying to access the "Recently added" movies and "Movies" menu items now fails to produce the expected result, and instead provides just a list of sources. There are errors in the [debug log](http://sprunge.us/PWfL) which would seem to be responsible for the issues - these errors are listed below.

When accessing "Recently added" movies menu:
```
22:05:21 152.282059 T:1966665728   DEBUG: Activating window ID: 10025
22:05:21 152.519897 T:1966665728   DEBUG: ------ Window Deinit (Home.xml) ------
22:05:21 152.564026 T:1966665728   DEBUG: ------ Window Init (MyVideoNav.xml) ------
22:05:21 152.567581 T:1966665728   DEBUG: CGUIMediaWindow::GetDirectory (videodb://recentlyaddedmovies/)
22:05:21 152.567780 T:1966665728   DEBUG:   ParentPath = [videodb://recentlyaddedmovies/]
22:05:21 152.587646 T:1966665728   DEBUG: RunQuery took 5 ms for 25 items query: select * from movie_view  ORDER BY dateAdded desc, idMovie desc LIMIT 25
22:05:21 152.588272 T:1966665728   ERROR: GetMoviesByWhere failed
22:05:21 152.589111 T:1966665728   ERROR: GetDirectory - Error getting videodb://recentlyaddedmovies/
22:05:21 152.590744 T:1966665728   ERROR: CGUIMediaWindow::GetDirectory(videodb://recentlyaddedmovies/) failed
22:05:21 152.590912 T:1966665728   DEBUG: CGUIMediaWindow::GetDirectory ()
22:05:21 152.591034 T:1966665728   DEBUG:   ParentPath = [videodb://recentlyaddedmovies/]
```

"Movies" menu:
```
22:06:58 249.738586 T:1966665728   DEBUG: Activating window ID: 10025
22:06:59 249.988220 T:1966665728   DEBUG: ------ Window Deinit (Home.xml) ------
22:06:59 250.041855 T:1966665728   DEBUG: ------ Window Init (MyVideoNav.xml) ------
22:06:59 250.044449 T:1966665728   DEBUG: CGUIMediaWindow::GetDirectory (videodb://movies/titles/)
22:06:59 250.044617 T:1966665728   DEBUG:   ParentPath = [videodb://movies/titles/]
22:06:59 250.210526 T:1966665728   DEBUG: RunQuery took 140 ms for 705 items query: select * from movie_view
22:06:59 250.211975 T:1966665728   ERROR: GetMoviesByWhere failed
22:06:59 250.225693 T:1966665728   ERROR: GetDirectory - Error getting videodb://movies/titles/
22:06:59 250.229416 T:1966665728   ERROR: CGUIMediaWindow::GetDirectory(videodb://movies/titles/) failed
22:06:59 250.229599 T:1966665728   DEBUG: CGUIMediaWindow::GetDirectory ()
22:06:59 250.229721 T:1966665728   DEBUG:   ParentPath = [videodb://movies/titles/]
```
It's the same for TV episodes.

When accessing the "Recently added" tvshows menu:
```
22:07:27 278.320251 T:1966665728   DEBUG: Activating window ID: 10025
22:07:27 278.570190 T:1966665728   DEBUG: ------ Window Deinit (Home.xml) ------
22:07:27 278.624664 T:1966665728   DEBUG: ------ Window Init (MyVideoNav.xml) ------
22:07:27 278.627197 T:1966665728   DEBUG: CGUIMediaWindow::GetDirectory (videodb://recentlyaddedepisodes/)
22:07:27 278.627380 T:1966665728   DEBUG:   ParentPath = [videodb://recentlyaddedepisodes/]
22:07:27 278.695038 T:1966665728   DEBUG: RunQuery took 42 ms for 25 items query: select * from episode_view  ORDER BY dateAdded desc, idEpisode desc LIMIT 25
22:07:27 278.695923 T:1966665728   ERROR: GetEpisodesByWhere failed
22:07:27 278.696747 T:1966665728   ERROR: GetDirectory - Error getting videodb://recentlyaddedepisodes/
22:07:27 278.699097 T:1966665728   ERROR: CGUIMediaWindow::GetDirectory(videodb://recentlyaddedepisodes/) failed
22:07:27 278.699249 T:1966665728   DEBUG: CGUIMediaWindow::GetDirectory ()
22:07:27 278.699341 T:1966665728   DEBUG:   ParentPath = [videodb://recentlyaddedepisodes/]
22:07:27 278.720856 T:1796207616  NOTICE: Thread BackgroundLoader start, auto delete: false
```

"TV Shows" menu:
```
22:08:22 333.064423 T:1966665728   DEBUG: Activating window ID: 10025
22:08:22 333.303375 T:1966665728   DEBUG: ------ Window Deinit (Home.xml) ------
22:08:22 333.350494 T:1966665728   DEBUG: ------ Window Init (MyVideoNav.xml) ------
22:08:22 333.353027 T:1966665728   DEBUG: CGUIMediaWindow::GetDirectory (videodb://tvshows/titles/)
22:08:22 333.353180 T:1966665728   DEBUG:   ParentPath = [videodb://tvshows/titles/]
22:08:22 333.379181 T:1966665728   DEBUG: RunQuery took 8 ms for 63 items query: SELECT * FROM tvshow_view
22:08:22 333.379730 T:1966665728   ERROR: GetTvShowsByWhere failed
22:08:22 333.380768 T:1966665728   ERROR: GetDirectory - Error getting videodb://tvshows/titles/
22:08:22 333.383057 T:1966665728   ERROR: CGUIMediaWindow::GetDirectory(videodb://tvshows/titles/) failed
22:08:22 333.383209 T:1966665728   DEBUG: CGUIMediaWindow::GetDirectory ()
22:08:22 333.383423 T:1966665728   DEBUG:   ParentPath = [videodb://tvshows/titles/]
```

Razzeee commented:
This is unexpected and I can't reproduce it with windows or I'm not understanding how you break it.

Also are you using sqlite in this example?

MilhouseVH commented:
No, MySQL.

phil65 commented:
just guessing here, perhaps it´s related to the typo in database update which was fixed later on?

Razzeee commented:
@phil65 
that should only have broken musicvideos

MilhouseVH commented:
I think I know what the problem is - this PR should be using v96 not v95.

#7920 already bumped the videodb version to v95 in master two days ago. Now anyone including this PR, who is already on v95, won't get the extra columns added and this causes the problems.

I've dropped my v95 db so that it v94 is migrated by this PR, and now there is no problem.

Razzeee commented:
@MilhouseVH 
yep, sorry. Seems like that got past me when rebasing this :(

MilhouseVH commented:
@Razzeee: Can you push a fix?

Razzeee commented:
I'm not sure if it's trivial. If you just increase the version and do the changes with 96, it will throw errors for every user that has these columns.

MilhouseVH commented:
Yes, but there's likely to be more v95 databases out there pre-dating this PR than those created by this PR. Adding a quick bump will only cause problems for the subset of users that created their v95 db with this PR.

Razzeee commented:
Then your code is fine. We will have to die one death here.

MilhouseVH commented:
Yeah, I've created the PR. Up to you guys to use it or not - I think it's the best solution, as I've got at least 200 users on v95 already that will have problems with this PR as it stands. Would be good to know what you want to do soon as it's tricky for me to release further test builds until this is resolved one way or the other (I can't include my PR unless it's merged because any decision not to merge will cause even more mayhem!)

MilhouseVH commented:
Would this PR be responsible for the issue reported [here](http://forum.kodi.tv/showthread.php?tid=231092&pid=2101425#pid2101425) which is affecting the YouTube addon? Based on the report, the last working build is [#0902](http://forum.kodi.tv/showthread.php?tid=231092&pid=2096442#pid2096442) which would mean the problem first appears in [#0904](http://forum.kodi.tv/showthread.php?tid=231092&pid=2098044#pid2098044) - the most likely change appears to be this PR.

Razzeee commented:
I think he just needs to change the sort order. Can't think of a reason this PR would cause that.
And I can't reproduce, other then setting the sort order to plays and having the same view.

MilhouseVH commented:
OK cheers, will suggest that to him.

uNiversaI commented:
After this PR was merged and Milhouses fix DB bump went in, I compiled Kodi:

I can tell you that I experienced some sort order issues temporarily.
It first presented itself as sorting being utterly wrong in skin (any list type view)
and changing sort order had no effect anywhere in skin, but after a while that went away, I changed sort order a few times to no effect, and watched something, when I came back enabled debug and try to submit the bug report the issue seems to have resolved itself.

Have tried youtube addons but... I will try and report back to http://forum.kodi.tv/showthread.php?tid=231092&pid=2101425#pid2101425

Razzeee commented:
That will need another PR which I was struggling some days ago with. Also skin integration, which I can't provide.
